source/common/stream_info/filter_state_impl.cc

```diff
@@ -5,8 +5,21 @@
 namespace Envoy {
 namespace StreamInfo {
 
-void FilterStateImpl::setData(absl::string_view data_name, std::unique_ptr<Object>&& data,
-                              FilterState::StateType state_type) {
+void FilterStateImpl::setData(absl::string_view data_name, std::shared_ptr<Object> data,
+                              FilterState::StateType state_type, FilterState::LifeSpan life_span) {
+  if (life_span > life_span_) {
+    if (hasDataWithNameInternally(data_name)) {
+      throw EnvoyException(
+          "FilterState::setData<T> called twice with conflicting life_span on the same data_name.");
+    }
+    maybeCreateParent(/* read_only = */ false);
```

Comment:
nit: if you find yourself wanting to have these comments here, I would recommend making this function take a dual state enum that will provide more clarity. TIOLO.